File: downbot.py
"""Provides a Discord bot redundancy system"""

import os
import json
import platform
import asyncio
import subprocess
import wget
import discord
from discord.ext.commands import Bot

CONFIG = {
    'discord_token': "Put Discord API Token here.",
    'id_to_watch': 0,
    'notify_id': [],
    'time_to_wait': 120,
    'directory': 'Put the COMPLETE path of the folder the bot is in here',
    'start_command': 'Put the command to start the bot here'
}
CONFIG_FILE = 'CONFIG.json'

if os.path.isfile(CONFIG_FILE):
    with open(CONFIG_FILE) as f:
        CONFIG.update(json.load(f))

with open('CONFIG.json', 'w') as f:
    json.dump(CONFIG, f, indent='\t')

print(platform.system())
START_COMMAND = CONFIG['start_command']
try:
    open('./startscript.sh', 'r')
except FileNotFoundError:
    wget.download('https://raw.githubusercontent.com/tweirtx/DownBot/scripts/startscript.sh')


bot = discord.ext.commands.Bot(command_prefix='#!')
loop = asyncio.AbstractEventLoop()
BOT = Bot(command_prefix='#!')


class StartBot():
    """Keeps track of certain variables"""
    cancelled = False
    handle = None
    in_alarm = False


async def start_up():
    """Starts the backup process"""
    await asyncio.sleep(CONFIG['time_to_wait'])
    if not StartBot.cancelled:
        StartBot.handle = subprocess.Popen(START_COMMAND, shell=True, cwd=CONFIG['directory'])
    if StartBot.cancelled:
        StartBot.cancelled = False


@BOT.event
async def on_member_update(before, after):
    """If the bot goes offline, trigger everything that should happen"""
    if before.id == CONFIG['id_to_watch']:
        if after.status == discord.Status('offline'):
            for i in CONFIG['notify_id']:
                person = before.guild.get_member(i)
                await person.send("NOTICE: {} has gone offline. Starting backup process in {} seconds. "
                                  "Resolve outage or send #!cancel to cancel.".format(before.display_name,
                                                                                      CONFIG['time_to_wait']))
            StartBot.cancelled = False
            await start_up()
        elif after.status == discord.Status('online') and before.status == discord.Status('offline'):
            StartBot.cancelled = True
        elif before.status == discord.Status.dnd and after.status == discord.Status.online:
            print("Primary back online registered")
            StartBot.cancelled = True
        else:
            print("after.status is {} type {} and before.status is {}".format(after.status, type(after.status),
                                                                              before.status))


@BOT.command()
async def cancel(ctx):
    """Manually cancel a startup"""
    if StartBot.in_alarm:
        StartBot.cancelled = True
        StartBot.in_alarm = False
        await ctx.send("start up cancelled")
    else:
        await ctx.send("No start up to cancel!")


@BOT.command()
async def shutdown(ctx):
    """Manually shut down the process"""
    StartBot.in_alarm = False
    StartBot.cancelled = False
    StartBot.handle.kill()
    await ctx.send("Shutdown successful")


@BOT.command()
async def gethandle(ctx):
    """Sends the handle of the process (debug only)"""
    await ctx.send(StartBot.handle)


@BOT.event
async def on_ready():
    """Announce that the bot is ready once it is"""
    print("Ready")

BOT.run(CONFIG['discord_token'])

File: downbot.js
const Discord = require('discord.js');
const client = new Discord.Client();
const config = require('./CONFIG.json');
const pm2 = require('pm2');

var in_alarm = false;
var running;

function start() {
    if (in_alarm) {
        pm2.start({
            name: "downbot-backup-process",
            script: config.start_command,
            cwd: config.directory
        });
        running = true;
    }
}

async function alertAlertedPeople(whoToSend, whatToSend) {
    var memberToSend = await client.fetchUser(whoToSend);
    await memberToSend.send(whatToSend).catch(console.error)
}


client.on('ready', () => {
    console.log("Ready!");
});

client.on('message', msg => {
    if(msg.content == "#!cancel") {
        if (in_alarm) {
            if (!running) {
                msg.reply("cancelled");
                in_alarm = false;
            }
        }
        else {
            msg.reply("no startup to cancel!")
        }
    }
    if(msg.content == "#!shutdown") {
        if (running) {
            pm2.stop("downbot-backup-process");
            in_alarm = false;
            running = false;
            msg.reply("sucessfully shut down");
        }
        else {
            msg.reply("cannot shut down something that is not running!!");
        }
    }
    if(msg.content == "#!ping") {
        msg.reply("Pong!");
    }
});

client.on('presenceUpdate', (oldMember, newMember) => {
    if (oldMember.id == config.id_to_watch) {
        if(oldMember.presence.status == "online") {
            if (newMember.presence.status == "offline") {
                in_alarm = true;
                console.log("Alarm detected");
                var memb;
                for (memb in config.notify_id) {
                    var membid = config.notify_id[memb];
                    var prometo = alertAlertedPeople(membid, "The bot this bot is responsible for monitoring has gone down. Starting automatically if #!cancel is not sent.")
                }
                setTimeout(start, config.time_to_wait * 1000);
            }
        }
        if (oldMember.presence.status == "dnd") {
            if (newMember.presence.status == "online") {
                console.log('stop detected');
                pm2.stop("downbot-backup-process");
                running = false;
                in_alarm = false;
            }
        }
    }
});

client.login(config.discord_token);